feat(sql): Implement `ST_Azimuth()`

yutannihilation opened the pull request:
This pull request is my attempt to implement `ST_Azimuth()`.

```
> sd_sql("select ST_Azimuth( ST_Point(25, 45),  ST_Point(75, 100))")
┌──────────────────────────────────────────────────────────────────────────┐
│ st_azimuth(st_point(Int64(25),Int64(45)),st_point(Int64(75),Int64(100))) │
│                                  float64                                 │
╞══════════════════════════════════════════════════════════════════════════╡
│                                                       0.7378150601204649 │
└──────────────────────────────────────────────────────────────────────────┘
Preview of up to 6 row(s)
```

This pull request is mostly for figuring out how to implement a function so that I can contribute more. So, please feel free to close if there's another plan to implement this.

I picked `ST_Azimuth()` because it looks relatively simple. I follow the implementation of Sedona, although I might not understand the code correctly as I'm poor at Java...

https://github.com/apache/sedona/blob/f9069e7dc6682d53335f0e0c6fb4bd444024d3b5/common/src/main/java/org/apache/sedona/common/Functions.java#L202-L209

I think the implementation is straightforward. One thing to note is that, if I understand the Sedona's implementation correctly, it returns `0.0` when the two points are the same. However, PostGIS returns `NULL`. [The document](https://postgis.net/docs/en/ST_Azimuth.html) says:

> NULL if the two points are coincident

petern48 commented on rust/sedona-geo/src/st_azimuth.rs:
```diff
@@ -0,0 +1,174 @@
+}
+
+// Note: When the two points are completely coincident, PostGIS's ST_Azimuth()
+//       returns NULL. However, this returns 0.0.
```
Let's definitely add a test for this case.

But first, we should confirm what the desired behavior is. I suspect this is a just something that was missed in the original Sedona, and maybe we should follow PostGIS behavior here instead. There doesn't seem to be any discussion about this in the [original Sedona PR](https://github.com/apache/sedona/pull/449). @jiayuasu was this difference intentional or maybe just something that was missed? Shall we fix it in Sedona as well?

petern48 commented on rust/sedona-functions/src/st_azimuth.rs:
```diff
@@ -0,0 +1,66 @@
+    .with_argument("geomA", "geometry: Start point geometry")
+    .with_argument("geomB", "geometry: End point geometry")
+    .with_sql_example(
+        "SELECT ST_Azimuth(\n            ST_Point(0, 0),\n            ST_Point(1, 1)\n        )",
```
```suggestion
        "SELECT ST_Azimuth(ST_Point(0, 0), ST_Point(1, 1))",
```
Let's have the example all on one line, like we how it is in the rest of the other functions.

petern48 commented:
Also, add a benchmark here in [native-functions.rs](https://github.com/apache/sedona-db/blob/7f911353335bd2fb51148aadb39a858a5ec12ffd/rust/sedona-functions/benches/native-functions.rs#L29) (they're in alphabetical order).

There's an example of a bench for a function with two geometries as inputs [here](https://github.com/apache/sedona-db/blob/7f911353335bd2fb51148aadb39a858a5ec12ffd/c/sedona-geos/benches/geos-functions.rs#L66-L72)

yutannihilation commented:
Thanks for reviewing!

yutannihilation commented:
I think I addressed all the comments. Thanks for the detailed explanation, it really helps!

I switch the implementation to return `NULL` for the coincident points case, but I'll wait for confirmation.

yutannihilation commented on rust/sedona-functions/src/st_azimuth.rs:
```diff
@@ -0,0 +1,221 @@
+                    end_coord.x(),
+                    end_coord.y(),
+                )),
+                // If either of the points is empty, the result is NULL
+                _ => Ok(None),
```
It seems PostGIS also rejects MULTIPOINT.

```sh
postgres=# SELECT ST_Azimuth(ST_Point(0, 0), ST_GeomFromText('MULTIPOINT (1 1)'));
ERROR:  Argument must be POINT geometries
```

yutannihilation commented:
During adding these tests, I found PostGIS raise errors for these two cases where my implementation returns `NULL`.

The first one is two `NULL`s. I think test should not fail with this error because `NULL::geometry` works. However, casting to geometry doesn't work on SedonaDB yet.

```
postgres=# SELECT ST_Azimuth(NULL, NULL);
ERROR:  function st_azimuth(unknown, unknown) is not unique
LINE 1: SELECT ST_Azimuth(NULL, NULL);
               ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
```

The second one is empty point. I thought this returns `NULL` without errors, but it seems PostGIS doesn't allow empty point. I think we can follow this behavior.

```
postgres=# SELECT ST_Azimuth(ST_Point(0, 0), ST_GeomFromText('POINT EMPTY'));
NOTICE:  lwgeom_api.c [351] called with n=0 and npoints=0
ERROR:  Error extracting point
```

yutannihilation commented:
Thanks for reviewing!

@jiayuasu 
I'm not familiar with benchmarking. Are there any guidance to do this? Do you mean I should have added a case here?

https://github.com/apache/sedona-db/blob/main/benchmarks/test_functions.py

jiayuasu commented:
@yutannihilation we have a tool in SedonaDB to produce benchmark results like what's shown in this PR: https://github.com/apache/sedona-db/pull/171

@petern48 @paleolimbot can you help @yutannihilation figure out how to run it?

jiayuasu commented:
I think you are right. Looking forward to it!

petern48 commented:
I'm not sure how you got actual `s`'s (skips), but I suspect it's because you tried running the entire file, which would take forever. I submitted a [PR](https://github.com/apache/sedona-db/pull/188) to do this myself, since it requires adding a new points-only table. It wouldn't have been fun for you to figure out and debug since our current setup just hangs for a while (minutes) and eventually outputs a sort of useful error message. Also updated the docs to mention running one bench at a time.

Here's the benchmark screenshot copied over.

<img width="697" height="95" alt="image" src="https://github.com/user-attachments/assets/603533ac-25d2-45ab-86a2-6725d24aeb41" />

> Since DuckDB's ST_Azimuth() is also implemented by me, you can blame me if DuckDB is not slow enough :) Anyway, looking forward to seeing the result!

I guess I'll instead thank you that DuckDB is slow enough 😉

petern48 commented:
Got it, thanks for catching that. It helps smooth the process for everyone when we know to document these small details. It's not ideal that the whole thing is skipped when the container isn't running, but I added a note about it to that same PR for now.
